app/com/lucidchart/open/cashy/controllers/UploadController.scala

```diff
@@ -0,0 +1,142 @@
+package com.lucidchart.open.cashy.controllers
+
+import com.lucidchart.open.cashy.views
+import com.lucidchart.open.cashy.models.{AssetModel, AuditModel}
+import com.lucidchart.open.cashy.request.{AppFlash, AuthAction}
+import com.lucidchart.open.cashy.amazons3.S3Client
+import com.lucidchart.open.cashy.utils.{GzipHelper, FileHandler}
+
+import java.io.File
+import play.api.Logger
+import play.api.mvc.MultipartFormData.FilePart
+import play.api.mvc.{Request, Action}
+import play.api.Play.current
+import play.api.Play.configuration
+import play.api.data._
+import play.api.data.Forms._
+import scala.collection.JavaConverters._
+import validation.Constraints
+
+case class UploadFailedException(message: String) extends Exception(message)
+
+object UploadController extends UploadController
+class UploadController extends AppController {
+
+  val logger = Logger(this.getClass)
+
+  private val bucketCloudfrontMap = configuration.getConfig("amazon.s3.bucketCloudfrontMap").get.subKeys.map(k => (k -> configuration.getString(s"amazon.s3.bucketCloudfrontMap.$k.cloudfront").get)).toMap
+  val buckets: Set[String] = bucketCloudfrontMap.keys.toSet
+  val uploadExtensions: List[String] = configuration.getStringList("upload.extensions").get.asScala.toList
+  val minNestedDirectories: Int = configuration.getInt("upload.minNestedDirectories").get
+  val minGzipSavings: Double = configuration.getDouble("upload.minGzipSavings").get
+
+  private case class UploadFormSubmission(
+    bucket: String,
+    assetName: String
+  )
+  private val uploadForm = Form(
+    mapping(
+      "bucket" -> text.verifying("Invalid bucket", x => buckets.contains(x)),
+      "assetName" -> text.verifying("Enter a name", x => !x.isEmpty)
+              .verifying("Must be organized in at least " + minNestedDirectories + " directories", x => x.split("/").length >= minNestedDirectories + 1)
+              .verifying("Must end in a valid extension (" + uploadExtensions.map("." + _).mkString(", ") + ")", x => checkExtension(x))
+    )(UploadFormSubmission.apply)(UploadFormSubmission.unapply) verifying("Name not available", form => form match {
```

Comment:
In order to verify that the assetName didn't already exist, and also using the value from the form, I had to verify on the entire form.  So the error message now shows up below the form for this particular error